journal/read.go

```diff
@@ -0,0 +1,191 @@
+	NumFromTail uint64        // start relative to the tail
+
+	// Show only journal entries whose fields match the supplied values. If
+	// the array is empty, entries will not be filtered.
+	Matches []Match
+
+	MaxEntryLength uint64 // the max allowable length for an entry (default is 64K)
+}
+
+// JournalReader is an io.ReadCloser which provides a simple interface for iterating through the
+// systemd journal.
+type JournalReader struct {
+	journal *Journal
+}
+
+// NewJournalReader creates a new JournalReader with configuration options that are similar to the
+// systemd journalctl tool's iteration and filtering features.
+func NewJournalReader(config JournalReaderConfig) (*JournalReader, error) {
+	var err error
+	r := &JournalReader{}
+
+	// Open the journal
+	if r.journal, err = NewJournal(); err != nil {
+		return nil, err
+	}
+
+	// Add any supplied matches
+	for _, m := range config.Matches {
+		r.journal.AddMatch(m.String())
+	}
+
+	// Set the start position based on options
+	if config.Since != 0 {
+		// Start based on a relative time
+		start := time.Now().Add(config.Since)
+		if err := r.journal.SeekRealtimeUsec(uint64(start.UnixNano() / 1000)); err != nil {
+			return nil, err
+		}
+	} else if config.NumFromTail != 0 {
+		// Start based on a number of lines before the tail
+		if err := r.journal.SeekTail(); err != nil {
+			return nil, err
+		}
+
+		// Move the read pointer into position near the tail. Go one further than
+		// the option so that the initial cursor advancement positions us at the
+		// correct starting point.
+		if _, err := r.journal.PreviousSkip(config.NumFromTail + 1); err != nil {
+			return nil, err
+		}
+	}
+
+	return r, nil
+}
+
+func (r *JournalReader) Read(b []byte) (int, error) {
+	var err error
+	var c int
+
+	// Advance the journal cursor
+	c, err = r.journal.Next()
+
+	// An unexpected error
+	if err != nil {
+		return 0, err
+	}
+
+	// EOF detection
+	if c == 0 {
+		return 0, io.EOF
+	}
+
+	// Build a message
+	var msg string
+	msg, err = r.buildMessage()
+
+	if err != nil {
+		return 0, err
+	}
+
+	// Copy and return the message
+	copy(b, []byte(msg))
+
+	return len(msg), nil
+}
+
+func (r *JournalReader) Close() error {
+	return r.journal.Close()
+}
+
+// Follow synchronously follows the JournalReader, writing each new journal entry to writer. The
+// follow will continue until a single time.Time is received on the until channel.
+func (r *JournalReader) Follow(until <-chan time.Time, writer io.Writer) error {
+	// Holds journal events to process. Tightly bounded for now unless there's a
+	// reason to unblock the journal watch routine more quickly.
+	events := make(chan int, 1)
+
+	// Start a poll routine to notify us of journal events
+	// TODO: the timeout logic here isn't very good.
+	pollDone := make(chan bool, 1)
+	go func() {
+	poll:
+		for {
+			select {
+			case <-pollDone:
+				break poll
+			default:
+				events <- r.journal.Wait(uint64(time.Duration(2) * time.Second))
```

Comment:
This doesn't seem safe to me, unless I'm missing something we can't safely enter the sd_journal C api from multiple goroutines on the same journal instance.  This is basically a continuation of my concern mentioned above about a lack of mutex serializing concurrent access to r.journal.cjournal.